### PR TITLE
feat: add ServiceProxyResponseQueue

### PR DESCRIPTION
**Pull request by daantimmer**
ServiceProxyResponseQueue is a generic extension to ServiceProxy that adds an easy to reuse scalable ResponseQueue.
ServiceProxy can only queue one request to send. However there are usecases where more than one requests to send can be requested. This class adds that generic functionality.

Usage example:

```cpp
//MyProto.proto:
//service MyProtoService{
//    option (service_id) = 1;
//    rpc Ping(Nothing) returns (Nothing) { option (method_id) = 1; }
//}
//service MyProtoServiceResponse{
//    option (service_id) = 2;
//    rpc Pong(Nothing) returns (Nothing) { option (method_id) = 1; }
//}

//MyProtoEcho.hpp:
class MyProtoEcho : MyProtoService
{
public:
    void Ping() override;

private:
    services::ServiceProxyResponseQueue<MyProtoServiceResponseProxy>::WithStorage<5> response;
};

//MyProtoEcho.cpp:
void MyProtoEcho::Ping()
{
    response.RequestSend([this]{
        response.Pong();
    });
    MethodDone();
}

```

### CODE REVIEW

File: protobuf/echo/Echo.hpp [hunk 1]
```diff
@@ -494,6 +517,42 @@ namespace services
...
+        , container{ container }
+    {}
+
+    template<class TServiceProxy>
+    void ServiceProxyResponseQueue<TServiceProxy>::RequestSend(infra::Function<void()> onRequestGranted)
+    {
+        if (container.full())
+        {
+            return;
+        }
```
Comment (richardapeters) [inline protobuf/echo/Echo.hpp:534]:
```suggestion
            return;
```

File: protobuf/echo/Echo.hpp [hunk 2]
```diff
@@ -207,6 +208,28 @@ namespace services
         infra::Function<void()> onGranted;
     };
 
+    template<class TServiceProxy>
```
Comment (rjaegers) [inline protobuf/echo/Echo.hpp:211]: TServiceProxy -> ServiceProxyType.

File: protobuf/echo/Echo.hpp [hunk 3]
```diff
@@ -207,6 +208,28 @@ namespace services
...
+        explicit ServiceProxyResponseQueue(Container& container, TArgs&&... args);
+
+        void RequestSend(infra::Function<void()> onRequestGranted) override;
+
+    private:
+        void ProcessSendQueue();
+
+        Container& container;
+
+        bool responseInProgress{ false };
```
Comment (rjaegers) [inline protobuf/echo/Echo.hpp:230]:
```suggestion
        void ProcessSendQueue();

    private:
        Container& container;
        bool responseInProgress{ false };
```

**Review by rjaegers — CHANGES_REQUESTED** [review]
Please address the comments as mentioned below.

In addition: please add tests for the new class, and add a description to the pull request about the use case and what the class solves.

File: protobuf/echo/Echo.hpp (hunk 1 above)
Comment (daantimmer) [reply protobuf/echo/Echo.hpp:534]: yep, sorry 'bout that.

File: protobuf/echo/Echo.hpp (hunk 3 above)
Comment (daantimmer) [reply protobuf/echo/Echo.hpp:230]: Are you sure? I thought we were going to adhere to the sonarlint warnings to not have consecutive same-access specifier?

Comment (rjaegers) [reply protobuf/echo/Echo.hpp:231]: Yes, I am sure. We are going to disable that Sonar warning. Our coding guidelines clearly describe that we use the access specifiers to distinguish different blocks of code and/or methods and variables.

Comment (daantimmer) [reply protobuf/echo/Echo.hpp:231]: (y)

**Review by rjaegers — APPROVED** [review]

**Pull request merged**